File: PathFinder.java
import java.nio.file.Path;
import java.util.*;
public class PathFinder {
    HashMap<String,Node> nodes;
    String[] stringNodes;

    public PathFinder(HashMap<String,Double> paths, String[] stringNodes){
        this.stringNodes = stringNodes;
        this.nodes = new  HashMap<String,Node>();

        for (int i = 0; i < stringNodes.length; i++) {
            this.nodes.put(stringNodes[i], new Node(stringNodes[i]));
        }
         
        String entryString;
        for (Map.Entry<String,Double> entry : paths.entrySet()){
            entryString = entry.getKey();
            this.nodes.get(entryString.charAt(0)+"").setAdj(entryString.charAt(3)+"",   entry.getValue());
        }

    }

    public void search(String starNode, String endNode){
        //到達距離を無限遠に初期化
        for (int i = 0; i < stringNodes.length; i++) {
            this.nodes.get(stringNodes[i]).setD(Double.POSITIVE_INFINITY);
        }
        
        //キューを初期化
        Queue<String> queue = new LinkedList<String>();
        queue.add(starNode);
        //currNodeを初期化
        String currNode = queue.peek();
        this.nodes.get(starNode).setD(0);
        //初期ノードのprevNodeを初期化
        this.nodes.get(currNode).setPrev(null);
        
        //探索開始
        while (queue.size() > 0){
            currNode = queue.remove();
            
            HashMap<String,Double> adjs = this.nodes.get(currNode).getAdj();

            for (Map.Entry<String,Double> entry : adjs.entrySet()){
                if (entry.getValue() != 0.0){
                    double tmpD = this.nodes.get(currNode).getD() + (double) this.nodes.get(currNode).getAdj().get(entry.getKey());
                    if (this.nodes.get(entry.getKey()).getD() > tmpD){
                        this.nodes.get(entry.getKey()).setD(tmpD);
                        this.nodes.get(entry.getKey()).setPrev(this.nodes.get(currNode));
                        this.nodes.get(entry.getKey()).setExplored(true);
                        queue.add(entry.getKey());
                    }
                }
               
            }
        }

        
        Node node = this.nodes.get(endNode);

        //バックトラックの開始
        LinkedList<String> s = new LinkedList<String>();
        s.add(endNode);
        
         while(node.getPrev() != null){
            node = node.getPrev();
            s.add(node.getName());
        }

        while(!s.isEmpty()) {
            System.out.print(s.remove(s.size()-1) + "->");

        }
        System.out.println();
        System.out.println(this.nodes.get(endNode).getD());


        
        
        


        
    }




    public static void main(String[] args) {
        int[][] D = { 
            {0, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 0}, 
            {3, 0, 0, 0, 0, 0, 0}, 
            {0, 1, 4, 0, 0, 0, 0}, 
            {0, 2, 1, 0, 0, 3, 0}, 
            {2, 6, 2, 0, 0, 0, 0}, 
            {0, 2, 0, 0, 0, 5, 0} 
        };

        int[][] DT = new int[D.length][D.length];

        for (int i = 0; i < DT.length; i++) {
            for (int j = 0; j < DT.length; j++) {
                    DT[i][j] = D[j][i];
            }
        }
        
        String[] nodeStrings = new String[D.length];
        for (int i = 0; i < nodeStrings.length; i++) {
            nodeStrings[i] = ((char) (i+97) +"").toUpperCase();
        }

        //全経路と重みの取得
        HashMap<String,Double> allPaths = new HashMap<String,Double>();

        double[][] G = new double[D.length][D.length];
        for (int i = 0; i < G.length; i++) {
            for (int j = 0; j < G.length; j++){
                if (i != j){
                    G[i][j] = D[i][j] + DT[i][j];
                    allPaths.put(nodeStrings[i] + "->" + nodeStrings[j],G[i][j]);
                    System.out.println(nodeStrings[i] + "->" + nodeStrings[j] + " " + G[i][j]);
                }
               
            }
            
        }


        PathFinder p = new PathFinder(allPaths, nodeStrings);
        p.search("A", "B");
    }


}

File: PathFinder.py
import random
import numpy as np
import matplotlib.pyplot as plt
import networkx as n 
from graphviz import Graph

class AdjNode:
    def __init__(self, weight: int, is_Barrier_free: bool) -> None:
        self.weight: int = weight
        self.is_Barrier_free: bool = is_Barrier_free

    def __add__(self, other):
        return self.weight + other.weight
    
    def __repr__(self) -> str:
        return f"{self.weight*1 if self.is_Barrier_free else self.weight*-1}"

def transpose(G:list[list[AdjNode]]) -> list[list[AdjNode]]:
    rows = len(G)
    cols = len(G[0])
    G_T = [[None for _ in range(rows)] for _ in range(cols)]
    for i in range(rows):
        for j in range(cols):
            G_T[j][i] = G[i][j]
    return G_T

def generate_random_G(R:int) -> np.ndarray:
    random.seed(10)
    weights_cs = [0]*50 + [1]*4 + [2]*2 + [3]*2 + [3]*6 + [4]*3 + [5]*6
    p = [-1]*5 + [1]*12
    def get_weight():
        return random.choice(weights_cs) * random.choice(p)
    G : list[AdjNode] = [[AdjNode(0,False) for i in range(R)] for j in range(R)]
    for i in range(R):
        for j in range(R):
            if i == j or i < j:
                continue
            w = get_weight()
            G[i][j].weight = abs(w)
            G[i][j].is_Barrier_free = True if w > 0 else False
    return  G

G = generate_random_G(10)

GraphvizObject = Graph(format="png")
GraphvizObject.attr("node", shape="circle")

# name nodes from A to A+R
G : list[AdjNode] = G + transpose(G)
nodes, path_weights = [str(chr(i).upper()) for i in range(97,len(G)+97)], {}

for i in range(len(G)):
    for j in range(len(G[0])):
        if G[i][j].weight != 0 :
            if i>j:
                print(type(G[i][j]))
                GraphvizObject.edge(str(nodes[i]),str(nodes[j]),label=str(G[i][j].weight),color='black' if G[i][j].is_Barrier_free else 'red')
            path_weights[f"{nodes[i]}->{nodes[j]}"] = G[i][j]
GraphvizObject.render("G")




print(path_weights)

class Node:
    def __init__(self,name:str) -> None:
        self.name:str = name
        self.adj:dict[str,int] = {}
        self.d:float = float('inf')
        self.explored:bool = False
        self.prevNode:Node | None = None
    
    def __repr__(self) -> str:
        return self.name

    def set_explored(self,e:bool) -> None:
        self.explored = e

    def set_d(self,d:int) -> None:
        self.d = d 



class Node:
    def __init__(self,name:str) -> None:
        self.name:str = name
        self.adj:dict[str,int] = {}
        self.d:float = float('inf')
        self.explored:bool = False
        self.prevNode:Node | None = None
    
    def __repr__(self) -> str:
        return self.name

    def set_explored(self,e:bool) -> None:
        self.explored = e

    def set_d(self,d:int) -> None:
        self.d = d 

class PathFinder:
    def __init__(self,all_paths:dict[str,int],nodes_string:list[str]) -> None:
        self.nodes_string: list[str] = nodes_string
        self.nodes:dict[str,Node] = {i:Node(i) for i in nodes_string}
        for i in all_paths.keys():
            self.nodes[i[0]].adj[i[3]] = all_paths[i]
    
    def dijkstra(self,startNode:str,endNode:str):

        #最短距離を無限大に初期化
        for i in self.nodes.keys():
            self.nodes[i].d = float('inf')

       
        queue:list[str] = [startNode]
        #currNodeを初期化
        currNode: str = queue[0]
        #currNodeの距離を初期化
        self.nodes[currNode].set_d(0)
        #初期ノードのprevNodeを初期化
        self.nodes[currNode].prevNode = None


        while len(queue) >0:
            currNode = queue.pop(0)
            adjs = self.nodes[currNode].adj
            for i in adjs.keys():
                tmp_d:float = self.nodes[currNode].d + self.nodes[currNode].adj[i]
                if self.nodes[i].d > tmp_d:
                    #スタートノードからの最短到達距離を更新
                    self.nodes[i].set_d(int(tmp_d))
                    self.nodes[i].prevNode = self.nodes[currNode]
                    self.nodes[i].set_explored(True)
                    queue.append(i)
        cNode:Node = self.nodes[endNode]
    
        #Backtrack
        s:list[str] = [endNode]
        while cNode.prevNode != None:
            cNode = cNode.prevNode
            s.append(cNode.name)

        return [s[::-1],self.nodes[endNode].d]

W = PathFinder(path_weights,nodes)

print(W.search_path("A","B"))

 

